src/main/java/com/icthh/xm/ms/entity/lep/transform/LepDataClassTransformation.java

```diff
@@ -0,0 +1,412 @@
+import org.codehaus.groovy.control.SourceUnit;
+import org.codehaus.groovy.transform.AbstractASTTransformation;
+import org.codehaus.groovy.transform.GroovyASTTransformation;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@GroovyASTTransformation(phase = CompilePhase.CANONICALIZATION)
+public class LepDataClassTransformation extends AbstractASTTransformation {
+
+    public static final String DOLLAR = "$";
+    public static final ConstantExpression NULL_VALUE = ConstantExpression.NULL;
+
+    private static final Logger log = LoggerFactory.getLogger(LepDataClassTransformation.class);
+    public static final VariableExpression THIS = varX("this");
+
+    public void visit(ASTNode[] nodes, SourceUnit source) {
+        try {
+            this.internalVisit(nodes);
+        } catch (Throwable e) {
+            log.error("Error during LepDataClassTransformation", e);
+            throw e;
+        }
+    }
+
+    private void internalVisit(ASTNode[] nodes) {
+
+        if (nodes == null || nodes.length == 0
+            || !(nodes[0] instanceof AnnotationNode annotation)
+            || !(nodes[1] instanceof ClassNode annotatedClass)
+            || annotatedClass.getFields() == null
+            || !annotation.getClassNode().getNameWithoutPackage().equals("LepDataClass")
+            || annotatedClass.isInterface()
+            || annotatedClass.isEnum()
+        ) {
+            return;
+        }
+
+        addMapConstructor(annotatedClass);
+        addToMapMethod(annotatedClass);
+    }
+
+    private void addToMapMethod(ClassNode annotatedClass) {
+        if (annotatedClass.hasMethod("toMap", new Parameter[0])) {
+            log.error("Class {} already has a method toMap", annotatedClass.getName());
+            throw new IllegalStateException("Class " + annotatedClass.getName() + " already has a method toMap");
+        }
+
+        MethodNode method = new MethodNode(
+            "toMap",
+            ACC_PUBLIC,
+            new ClassNode(Map.class),
+            new Parameter[0],
+            ClassNode.EMPTY_ARRAY,
+            new BlockStatement()
+        );
+
+        BlockStatement body = (BlockStatement) method.getCode();
+        body.addStatement(
+            declS(varX("result"), ctorX(new ClassNode(LinkedHashMap.class)))
+        );
+
+        for (FieldNode field : annotatedClass.getFields()) {
+            if (field.isStatic() || field.getName().contains(DOLLAR) || (field.getModifiers() & ACC_TRANSIENT) != 0 || isIgnored(field)) {
+                continue;
+            }
+
+            String fieldName = field.getName();
+            body.addStatement(
+                ifS(notX(isNullX(fieldX(fieldName))),
+                    assignS(
+                        propX(varX("result"), fieldName),
+                        transformFieldInToMap(field.getType(), fieldX(fieldName), 0)
+                    )
+                )
+            );
+        }
```

Comment:
can we use `annotatedClass.getFields().stream()` here & add `field.isStatic() || field.getName().contains(DOLLAR) || (field.getModifiers() & ACC_TRANSIENT) != 0 || isIgnored(field)` to the filter() ?

this part could be also moved to the method like addBodyStatement or smth like that:
`String fieldName = field.getName();
            body.addStatement(
                ifS(notX(isNullX(fieldX(fieldName))),
                    assignS(
                        propX(varX("result"), fieldName),
                        transformFieldInToMap(field.getType(), fieldX(fieldName), 0)
                    )
                )
            );`